Support configurable split size for metadata tables

jun-he opened the pull request:
Related issue https://github.com/apache/incubator-iceberg/issues/817

rdblue commented:
The table's split size shouldn't be used for metadata tables. I think we should just rely on the options passed to the scan. We will also need tests for this.

jun-he commented:
@rdblue  can you please take a look? Thanks.

rdblue commented on core/src/test/java/org/apache/iceberg/TestSplitPlanning.java:
```diff
@@ -147,18 +178,23 @@ private void appendFiles(Iterable<DataFile> files) {
     appendFiles.commit();
   }
 
+
```
Nit: unnecessary empty line.

rdblue commented on core/src/main/java/org/apache/iceberg/BaseTableScan.java:
```diff
@@ -225,6 +225,8 @@ public Expression filter() {
     if (options.containsKey(TableProperties.SPLIT_SIZE)) {
       splitSize = Long.parseLong(options.get(TableProperties.SPLIT_SIZE));
+    } else if (options.containsKey(TableProperties.METADATA_SPLIT_SIZE)) {
+      splitSize = Long.parseLong(options.get(TableProperties.METADATA_SPLIT_SIZE));
```
I don't think this is necessary. When options are used to set the split size in the Spark reader, it uses `TableProperties.SPLIT_SIZE` to pass it here. That should work for metadata tables as well, right?

The situation that we need to handle in this PR is setting the default, like you had before. We just want to use a metadata split size so you can set it and not have the same split size used (by default) for both metadata and data scans.

jun-he replied:
@rdblue Thanks for the comments and updated the PR accordingly.

rdblue commented:
Thanks, @jun-he! It looks close, just need to combine the first version with this one.

rdblue commented on core/src/main/java/org/apache/iceberg/TableProperties.java:
```diff
@@ -106,6 +106,8 @@ private TableProperties() {}
   public static final String DEFAULT_WRITE_METRICS_MODE = "write.metadata.metrics.default";
   public static final String DEFAULT_WRITE_METRICS_MODE_DEFAULT = "truncate(16)";
 
+  public static final String METADATA_SPLIT_SIZE = "read.metadata.split-size";
```
Can we move this up by the other split size property? Also, the other name is `read.split.target-size` so let's use the same prefix. How about changing this to `read.split.metadata-target-size`?

rdblue commented:
Looks good other than the property name. Thanks @jun-he!

jun-he commented:
@rdblue Thanks for the review! Updated the PR accordingly.

jun-he commented:
@aokolnychyi  Thanks for the comments.
I am not sure what's the original reason for hard coded default value. I didn't change it previously as different metadata tables might set different default values. By checking the code, actually they are all the same (32MB). So I moved them into `TableProperties`. 
Also, I added additional tests in `TestDataSourceOptions`. Please take a look. Thanks.

aokolnychyi commented on spark/src/test/java/org/apache/iceberg/spark/source/TestDataSourceOptions.java:
```diff
@@ -284,4 +284,103 @@ public void testIncrementalScanOptions() throws IOException {
+        .mode("append")
+        .save(tableLocation);
+
+    int expectedSplits = ((int) tables.load(tableLocation + "#entries")
```
nit: I think it would be a bit more readable if we split this logic into multiple lines in all such places. However, it does not matter that much.

```
Table table = tables.create(...);

table.refresh();

List<ManifestFile> manifests = table.currentSnapshot().manifests();
int expectedSplitNum = (manifests.get(0).length() + splitSize -1) / splitSize;
```

jun-he commented:
@aokolnychyi  Thanks for the review. Updated the doc accordingly in #883.
